Defibrillator rewrite

mwerezak opened the pull request:
Technically a refactor, practically a rewrite though.

In short, I am surprised that defibrillators work at all given the state they are in.

Need to:

- [x] rewrite resuscitation so it's not as much of an awful noodley mess. 
- [x] rewrite heart and brain checks so that they aren't strewn about the file and not hardcoded
- [x] make brain damage due to dead time scale properly
- [x] adapt paddles to our twohanding code
- [x] use the correct power cell procs when draining charge.
- [x] charge handling sanity - *Almost. Still need to readd charge usage to the disarm and harm-intent usages of the defibrillator*
- [x] fix using defibrillators to revive synthetics.
- [x] replace `hurt_people()` behaviour with `electrocute_act`, less snowflake that way. 
- [x] replace `log_game()` (is that even used on baycode? It's also being called with strange arguments) with the appropriate admin messaging procs
- [x] ensure destroyed paddles are handled correctly.
- [x] unify `shockpaddles/proc/recharge()` and `defibrillator/proc/cooldowncheck()` which do basically the same thing except for `sleep` instead of `spawn` and cell charge deduction which shouldn't even be there anyways.
- [x] clean up cooldown handling so that it is clear how it is supposed to be delegated to the base unit
- [x] see if attacking with paddles needs to be adapted to our item attack procs (answer: it doesn't)
- [x] clean up `if(req_defib) defib.var_access`. ~~If `req_defib` guarantees that defib is not null then it should be sufficient to replace these with `if(defib) defib.stuff` (and much safer that way).~~ Ended up separating standalone paddles and ones linked to a base unit into different subtypes.
- [x] change adjustOxyloss() to setOxyLoss() when reviving people so that they aren't as likely to immediately die.
- [x] double check that paddles cannot get separated from the base unit
- [ ] ~~make brain damage accumulate in mob/Life() after death instead of being dealt to the player by the shock paddles.~~ Later PR
- [x] check core functionality

Techhead0 commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -616,6 +611,19 @@
+
+	var/brain_damage = Clamp((deadtime - DEFIB_TIME_LOSS)/(DEFIB_TIME_LIMIT - DEFIB_TIME_LOSS)*brain.max_damage, 0, brain.max_damage)
+
+	if(H.getBrainLoss() < brain_damage)
```
Why not just use `H.getBrainLoss()` in the `Clamp()` instead of 0?

STEALTH EDIT: Too sleepy to read more, so this is the only comment for now.

mwerezak replied:
Good idea.

mwerezak commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -523,99 +538,122 @@
-						H.adjustOxyLoss(H.health - halfwaycritdeath, 0)
-					else
-						var/overall_damage = H.getBruteLoss() + H.getFireLoss() + H.getToxLoss() + H.getOxyLoss()
-						H.adjustOxyLoss((H.health - halfwaycritdeath) * (H.getOxyLoss() / overall_damage), 0)
```
Could not make any sense of what this else clause was trying to accomplish, got rid of it.

Nero-07 replied:
Thats the part that heals some oxy damage so people with 200+ of it have a chane to live at least a few seconds

mwerezak replied:
That's what L588 does, the comment was specifically for L591.

mwerezak committed:
Defibrilator fixes, cleanup, improvements

Fixes defibs giving victims enough brain damage to kill them before
the time limit is reached
Fixes defibs potentially being usable to cure brain damage

Refactors some sphagetti code

Removes frankenstein /obj/item/weapon/twohanded from tg

Condense brain damage calculation

Cleans up defibrillator charge handling

Still need to update shock_people() and hurt_people(),
charge usage for those is commented out in the meantime

Removes req_defib. If a paddles item spawns on its own it is assumed to be standalone

Replaces hurt_people() behaviour with eletrocute_act and stun_effect_act

Also removes tg-style stunbatonesque disarm option

Prevents defibs from being used to revive synthetics or species with NO_SCAN

Replaces log_game() calls with logging procs

Cleans up cooldown handling

Now entirely the responsiblity of the paddles

Formats defibrilator announcements to look like spoken text

Make alive sets stat to UNCONSCIOUS

Reimplements safety var on paddles, moves emp_act and emag_act to paddles

Ensures that destroyed paddles are handled correctly

Cleans up paddle inventory handling, removes useless on var and unused proc

Reorganizes vars

Defibs come with highcap cells by default, adjusts charge cost

Reorganizes linked and standalone paddles into separate subtypes

Ensures reattach_paddles() is called with user

Fixes defibs having ui_action_click() but no action icon enabled

Adjusts defib announcement format

Less appearance churn, fix check_charge condition

Makes revive oxyloss try to put the patient on the edge of crit if possible

Clicking with empty hand

Removes delay when harm-shocking so that it might actually be useful

Makes the start message bolded and more obvious in exchange, since the action is much quicker

Moves defibrillator sprites to their own dmi

Adds EMP effect for standalone paddles

Shock paddle harm-intent behaviour

Now applied as a hit effect instead of overriding attack
Tweaks messages
Special combat defibs have faster chargeup time

Removes redundant key_name() in log message

Nero-07 commented:
> In short, I am surprised that defibrillators work at all given the state they are in.

I resent that.

It's probably true, but I resent it anyway.

Nero-07 commented:
> clean up if(req_defib) defib.var_access. If req_defib guarantees that defib is not null then it should be sufficient to replace these with if(defib) defib.stuff (and much safer that way)

req defib is actually a config option. If you unset it, you need just the paddes and not the actual defib unit.

> change adjustOxyloss() to setOxyLoss() when reviving people so that they aren't as likely to immediately die.

I thought this was already the case. Feel free to change if it isn'T though

Minijar commented:
Could you expand the defib_time_limit to ten minutes? As that's officially how long the brain takes to start dying irl.

Nero-07 commented:
That seems very long. When I did the original port, people claimed anywhere from 5-8 minutes. The brain can lie there with no oxygen for 10 minutes before it even starts dying?

mustafakalash commented:
That's not let real life dictate what is balanced and enjoyable

mwerezak commented:
> I resent that. It's probably true, but I resent it anyway.

I was feeling grouchy and mean, and that was probably uncalled for. My apologies for that.

> req defib is actually a config option. If you unset it, you need just the paddes and not the actual defib unit.

I'm guessing that by config option you don't mean a server wide config, but rather a variable that is set differently for different paddle types and not expected to change over the course of the game? Either way it looks like it only needs to be checked at paddle creation time, not anywhere else.

> Could you expand the defib_time_limit to ten minutes?

~~Sure why not.~~

Raptor1628 commented:
Let's not tweak the brain survival rate quite yet. I'd like to have more testing before that's changed.

mwerezak commented:
For a moment there I was puzzled about whether the charge in `bcell` actually matters at all, then I noticed that the code which handles that was actually hidden away in `update_icon()` and distanced from the actual "can I use this" code by indirection through a state variable.

mwerezak commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -461,169 +348,221 @@
-	var/mob/living/carbon/human/H = M
+
 	if(user.a_intent == I_DISARM) //Shock them
 		shock_people(user, M)
```
Thinking of removing this. I don't see why defibrillators need to double as improvised stun batons, and it saves some work.

Nero-07 replied:
It's only after you emagged it. Feels natural that you could hurt people quite a bit with it *shrug

mwerezak replied:
Yes, and there is a proc for that, `hurt_people()` which will stay around in some form.

`shock_people()` is literally just click on someone to stun them.

mwerezak replied:
Yes, and there is a proc for that, `hurt_people()` which will stay around in some form.

`shock_people()` is literally just click on someone to stun them.

mwerezak commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -110,187 +104,60 @@
+
+	if((slot_flags & SLOT_BACK) && M.get_equipped_item(slot_back) == src)
+		return 1
+	if((slot_flags & SLOT_BELT) && M.get_equipped_item(slot_belt) == src)
```
I wonder why we don't do this more often for checking if an item is equipped to a slot? Might even someday lead to a proper inventory system that isn't just a bunch of vars on mob types.

mwerezak commented:
Standalone shockpaddles are now a separate subtype, instead of being mixed in together with code switching based on vars.

Added a desc inspired by the advanced e-gun and gives off a small amount of radiation when used (currently 10)

mwerezak commented:
Okay, should be good to go.

mwerezak commented:
Code-wise, the paddle item now does the bulk of the work, while the base unit mostly just stores the paddles, holds the power cell, and handles power cell/paddle removal/replacement.

The UI action button now actually appears.

Many other improvements...

Kasuobes commented:
Torch stalled again, poke it with a big stick.

PsiOmegaDelta commented on code/__defines/__unit_test.dm:
```diff
@@ -0,0 +1 @@
+#define UNIT_TEST
```
I assume this addition isn't intended.

mwerezak replied:
Nope

PsiOmegaDelta commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -301,329 +171,366 @@
+	
+	make_alive(H)
+
+	log_and_message_admins("[key_name(user)] used \a [src] to revive [key_name(H)].")
```
This should simply be ```log_and_message_admins("used \a [src] to revive [key_name(H)].")```, as name/key is already acquired from usr and applied first.
If you want to supply user specifically it's ```log_and_message_admins("used \a [src] to revive [key_name(H)].", user)```

mwerezak replied:
Ah, cool

mwerezak commented:
Updated the shock paddle harm-intent behaviour. Since the deliberate placement delay was removed, you have to hit the target in combat if they are standing, and the shock is applied as a hit effect. If you connect, the regular 2 second charge-up delay applies, but if the target doesn't move away after that time they receive a powerful stun. Combat and traitor paddles have a 1 second charge-up delay.

The harm-intent behaviour now happens regardless of safety, but fails early with a message if safety has a true value. This is intended to make the harm-intent behaviour more easily discoverable by players. It also means that slapping defibs on people doesn't give away that your unit is emagged.

The cooldown time is now only set if you get to the point where the paddles actually energize.

PsiOmegaDelta commented:
Didn't see any more glaring issues but is this a 31-commit worthy PR?

mwerezak commented:
Probably not, squash incoming.
